gcloud-java-pubsub/src/main/java/com/google/gcloud/pubsub/Message.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.gcloud.pubsub;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.util.Strings;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Pub/Sub message.
+ */
+public final class Message implements Serializable {
+
+  private static final long serialVersionUID = -1436515787233340634L;
+  private static final long NANOS_PER_MILLISECOND = 1000000;
+  private static final long MILLIS_PER_SECOND = 1000;
+
+  private final String id;
+  private final ByteString payload;
+  private final ImmutableMap<String, String> attributes;
+  private final Long publishTime;
+
+  /**
+   * Builder for Message.
+   */
+  public static final class Builder {
+
+    private String id;
+    private byte[] payload;
+    private Map<String, String> attributes = new HashMap<>();
+    private Long publishTime;
+
+    private Builder() {}
+
+    private Builder(Message message) {
+      id = message.id;
+      payload = message.payload.toByteArray();
+      attributes = new HashMap<>(message.attributes);
+      publishTime = message.publishTime;
+    }
+
+    Builder id(String id) {
+      this.id = id;
+      return this;
+    }
+
+    public Builder payload(byte[] payload) {
+      this.payload = checkNotNull(payload);
+      return this;
+    }
+
+    public Builder addAttribute(String name, String value) {
```

Comment:
This is not a gRPC communication settings but rather an add-on set of user-defined attributes to its message (see [here](https://cloud.google.com/pubsub/reference/rest/v1/PubsubMessage)).